### ISSUE TITLE
npm update fails when serialport was built manually

### ISSUE THREAD

**Issue by infusion** [EmergingTechnologyAdvisors/node-serialport#843]
Operating System and Hardware: Raspberry PI B+
NodeJS Version: v4.2.1
Serialport version:2.14.7
Python Version : Python 2.7.9

As said earlier at #833, I installed serialport on Raspberry PI and as there are no pre-built ARM packages, they were compiled manually by npm. Everything works fine with serialport, but another package was updated so I ran `npm update` and it failed for serialport now.

When I delete `node_modules` and run `npm install` again, it works as usual.

```
➜  ~ npm update
npm WARN engine serialport@1.2.5: wanted: {"node":"0.10.x"} (current: {"node":"4.2.1","npm":"2.14.7"})
-
> serialport@1.2.5 install /home/pi/node_modules/gps/node_modules/serialport
> node-gyp rebuild

make: Entering directory '/home/pi/node_modules/gps/node_modules/serialport/build'
  CXX(target) Release/obj.target/serialport/src/serialport.o
In file included from ../src/serialport.cpp:3:0:
../src/serialport.h:30:38: error: ‘Arguments’ in namespace ‘v8’ does not name a type
 v8::Handle<v8::Value> List(const v8::Arguments& args);
                                      ^
../src/serialport.h:31:15: error: variable or field ‘EIO_List’ declared void
 void EIO_List(uv_work_t* req);
               ^
../src/serialport.h:31:15: error: ‘uv_work_t’ was not declared in this scope
../src/serialport.h:31:26: error: ‘req’ was not declared in this scope
 void EIO_List(uv_work_t* req);
                          ^
../src/serialport.h:32:20: error: variable or field ‘EIO_AfterList’ declared void
 void EIO_AfterList(uv_work_t* req);
                    ^
../src/serialport.h:32:20: error: ‘uv_work_t’ was not declared in this scope
../src/serialport.h:32:31: error: ‘req’ was not declared in this scope
 void EIO_AfterList(uv_work_t* req);
                               ^
../src/serialport.h:34:38: error: ‘Arguments’ in namespace ‘v8’ does not name a type
 v8::Handle<v8::Value> Open(const v8::Arguments& args);
                                      ^
../src/serialport.h:35:15: error: variable or field ‘EIO_Open’ declared void
 void EIO_Open(uv_work_t* req);
               ^
../src/serialport.h:35:15: error: ‘uv_work_t’ was not declared in this scope
../src/serialport.h:35:26: error: ‘req’ was not declared in this scope
 void EIO_Open(uv_work_t* req);
                          ^
../src/serialport.h:36:20: error: variable or field ‘EIO_AfterOpen’ declared void
 void EIO_AfterOpen(uv_work_t* req);
                    ^
../src/serialport.h:36:20: error: ‘uv_work_t’ was not declared in this scope
../src/serialport.h:36:31: error: ‘req’ was not declared in this scope
 void EIO_AfterOpen(uv_work_t* req);
                               ^
In file included from ../src/serialport.cpp:3:0:
../src/serialport.h:39:39: error: ‘Arguments’ in namespace ‘v8’ does not name a type
 v8::Handle<v8::Value> Write(const v8::Arguments& args);
                                       ^
../src/serialport.h:40:16: error: variable or field ‘EIO_Write’ declared void
 void EIO_Write(uv_work_t* req);
                ^
../src/serialport.h:40:16: error: ‘uv_work_t’ was not declared in this scope
../src/serialport.h:40:27: error: ‘req’ was not declared in this scope
 void EIO_Write(uv_work_t* req);
                           ^
../src/serialport.h:41:21: error: variable or field ‘EIO_AfterWrite’ declared void
 void EIO_AfterWrite(uv_work_t* req);
                     ^
../src/serialport.h:41:21: error: ‘uv_work_t’ was not declared in this scope
../src/serialport.h:41:32: error: ‘req’ was not declared in this scope
 void EIO_AfterWrite(uv_work_t* req);
                                ^
../src/serialport.h:43:39: error: ‘Arguments’ in namespace ‘v8’ does not name a type
 v8::Handle<v8::Value> Close(const v8::Arguments& args);
                                       ^
../src/serialport.h:44:16: error: variable or field ‘EIO_Close’ declared void
 void EIO_Close(uv_work_t* req);
                ^
../src/serialport.h:44:16: error: ‘uv_work_t’ was not declared in this scope
../src/serialport.h:44:27: error: ‘req’ was not declared in this scope
 void EIO_Close(uv_work_t* req);
                           ^
../src/serialport.h:45:21: error: variable or field ‘EIO_AfterClose’ declared void
 void EIO_AfterClose(uv_work_t* req);
                     ^
../src/serialport.h:45:21: error: ‘uv_work_t’ was not declared in this scope
../src/serialport.h:45:32: error: ‘req’ was not declared in this scope
 void EIO_AfterClose(uv_work_t* req);
                                ^
../src/serialport.h:47:39: error: ‘Arguments’ in namespace ‘v8’ does not name a type
 v8::Handle<v8::Value> Flush(const v8::Arguments& args);
                                       ^
../src/serialport.h:48:16: error: variable or field ‘EIO_Flush’ declared void
 void EIO_Flush(uv_work_t* req);
                ^
../src/serialport.h:48:16: error: ‘uv_work_t’ was not declared in this scope
../src/serialport.h:48:27: error: ‘req’ was not declared in this scope
 void EIO_Flush(uv_work_t* req);
                           ^
../src/serialport.h:49:21: error: variable or field ‘EIO_AfterFlush’ declared void
 void EIO_AfterFlush(uv_work_t* req);
                     ^
../src/serialport.h:49:21: error: ‘uv_work_t’ was not declared in this scope
../src/serialport.h:49:32: error: ‘req’ was not declared in this scope
 void EIO_AfterFlush(uv_work_t* req);
                                ^
../src/serialport.h:89:3: error: ‘uv_work_t’ does not name a type
   uv_work_t req;
   ^
../src/serialport.h:90:3: error: ‘ngx_queue_t’ does not name a type
   ngx_queue_t queue;
   ^
In file included from ../src/serialport.cpp:8:0:
../src/serialport_poller.h:10:50: error: expected class-name before ‘{’ token
 class SerialportPoller : public node::ObjectWrap {
                                                  ^
../src/serialport_poller.h:23:46: error: ‘Arguments’ in namespace ‘v8’ does not name a type
   static v8::Handle<v8::Value> New(const v8::Arguments& args);
                                              ^
../src/serialport_poller.h:24:48: error: ‘Arguments’ in namespace ‘v8’ does not name a type
   static v8::Handle<v8::Value> Close(const v8::Arguments& args);
                                                ^
../src/serialport_poller.h:25:48: error: ‘Arguments’ in namespace ‘v8’ does not name a type
   static v8::Handle<v8::Value> Start(const v8::Arguments& args);
                                                ^
../src/serialport_poller.h:27:3: error: ‘uv_poll_t’ does not name a type
   uv_poll_t poll_handle_;
   ^
../src/serialport.cpp:11:1: error: ‘uv_mutex_t’ does not name a type
 uv_mutex_t write_queue_mutex;
 ^
../src/serialport.cpp:12:1: error: ‘ngx_queue_t’ does not name a type
 ngx_queue_t write_queue;
 ^
../src/serialport.cpp:14:38: error: ‘Arguments’ in namespace ‘v8’ does not name a type
 v8::Handle<v8::Value> Open(const v8::Arguments& args) {
                                      ^
In file included from /home/pi/.node-gyp/4.2.1/include/node/node.h:42:0,
                 from ../src/serialport.h:5,
                 from ../src/serialport.cpp:3:
/home/pi/.node-gyp/4.2.1/include/node/v8.h: In function ‘v8::Handle<v8::Value> Open(const int&)’:
/home/pi/.node-gyp/4.2.1/include/node/v8.h:885:13: error: ‘v8::HandleScope::HandleScope()’ is protected
   V8_INLINE HandleScope() {}
             ^
../src/serialport.cpp:15:19: error: within this context
   v8::HandleScope scope;
                   ^
../src/serialport.cpp:17:18: error: ‘write_queue_mutex’ was not declared in this scope
   uv_mutex_init(&write_queue_mutex);
                  ^
../src/serialport.cpp:17:35: error: ‘uv_mutex_init’ was not declared in this scope
   uv_mutex_init(&write_queue_mutex);
                                   ^
../src/serialport.cpp:18:19: error: ‘write_queue’ was not declared in this scope
   ngx_queue_init(&write_queue);
                   ^
../src/serialport.cpp:18:30: error: ‘ngx_queue_init’ was not declared in this scope
   ngx_queue_init(&write_queue);
                              ^
../src/serialport.cpp:21:13: error: invalid types ‘const int[int]’ for array subscript
   if(!args[0]->IsString()) {
             ^
../src/serialport.cpp:22:18: error: ‘class v8::HandleScope’ has no member named ‘Close’
     return scope.Close(v8::ThrowException(v8::Exception::TypeError(v8::String::New("First argument must be a string"))));
                  ^
../src/serialport.cpp:22:24: error: ‘ThrowException’ is not a member of ‘v8’
     return scope.Close(v8::ThrowException(v8::Exception::TypeError(v8::String::New("First argument must be a string"))));
                        ^
../src/serialport.cpp:22:68: error: ‘New’ is not a member of ‘v8::String’
     return scope.Close(v8::ThrowException(v8::Exception::TypeError(v8::String::New("First argument must be a string"))));
                                                                    ^
../src/serialport.cpp:24:36: error: invalid types ‘const int[int]’ for array subscript
   v8::String::Utf8Value path(args[0]->ToString());
                                    ^
../src/serialport.cpp:27:13: error: invalid types ‘const int[int]’ for array subscript
   if(!args[1]->IsObject()) {
             ^
../src/serialport.cpp:28:18: error: ‘class v8::HandleScope’ has no member named ‘Close’
     return scope.Close(v8::ThrowException(v8::Exception::TypeError(v8::String::New("Second argument must be an object"))));
                  ^
../src/serialport.cpp:28:24: error: ‘ThrowException’ is not a member of ‘v8’
     return scope.Close(v8::ThrowException(v8::Exception::TypeError(v8::String::New("Second argument must be an object"))));
                        ^
../src/serialport.cpp:28:68: error: ‘New’ is not a member of ‘v8::String’
     return scope.Close(v8::ThrowException(v8::Exception::TypeError(v8::String::New("Second argument must be an object"))));
                                                                    ^
../src/serialport.cpp:30:41: error: invalid types ‘const int[int]’ for array subscript
   v8::Local<v8::Object> options = args[1]->ToObject();
                                         ^
../src/serialport.cpp:33:13: error: invalid types ‘const int[int]’ for array subscript
   if(!args[2]->IsFunction()) {
             ^
../src/serialport.cpp:34:18: error: ‘class v8::HandleScope’ has no member named ‘Close’
     return scope.Close(v8::ThrowException(v8::Exception::TypeError(v8::String::New("Third argument must be a function"))));
                  ^
../src/serialport.cpp:34:24: error: ‘ThrowException’ is not a member of ‘v8’
     return scope.Close(v8::ThrowException(v8::Exception::TypeError(v8::String::New("Third argument must be a function"))));
                        ^
../src/serialport.cpp:34:68: error: ‘New’ is not a member of ‘v8::String’
     return scope.Close(v8::ThrowException(v8::Exception::TypeError(v8::String::New("Third argument must be a function"))));
                                                                    ^
../src/serialport.cpp:36:41: error: invalid types ‘const int[int]’ for array subscript
   v8::Local<v8::Value> callback = args[2];
                                         ^
../src/serialport.cpp:41:34: error: ‘New’ is not a member of ‘v8::String’
   baton->baudRate = options->Get(v8::String::New("baudRate"))->ToInt32()->Int32Value();
                                  ^
../src/serialport.cpp:42:34: error: ‘New’ is not a member of ‘v8::String’
   baton->dataBits = options->Get(v8::String::New("dataBits"))->ToInt32()->Int32Value();
                                  ^
../src/serialport.cpp:43:36: error: ‘New’ is not a member of ‘v8::String’
   baton->bufferSize = options->Get(v8::String::New("bufferSize"))->ToInt32()->Int32Value();
                                    ^
../src/serialport.cpp:44:45: error: ‘New’ is not a member of ‘v8::String’
   baton->parity = ToParityEnum(options->Get(v8::String::New("parity"))->ToString());
                                             ^
../src/serialport.cpp:45:48: error: ‘New’ is not a member of ‘v8::String’
   baton->stopBits = ToStopBitEnum(options->Get(v8::String::New("stopBits"))->ToNumber()->NumberValue());
                                                ^
../src/serialport.cpp:46:32: error: ‘New’ is not a member of ‘v8::String’
   baton->rtscts = options->Get(v8::String::New("rtscts"))->ToBoolean()->BooleanValue();
                                ^
../src/serialport.cpp:47:29: error: ‘New’ is not a member of ‘v8::String’
   baton->xon = options->Get(v8::String::New("xon"))->ToBoolean()->BooleanValue();
                             ^
../src/serialport.cpp:48:30: error: ‘New’ is not a member of ‘v8::String’
   baton->xoff = options->Get(v8::String::New("xoff"))->ToBoolean()->BooleanValue();
                              ^
../src/serialport.cpp:49:30: error: ‘New’ is not a member of ‘v8::String’
   baton->xany = options->Get(v8::String::New("xany"))->ToBoolean()->BooleanValue();
                              ^
../src/serialport.cpp:52:60: error: no matching function for call to ‘v8::Persistent<v8::Value>::New(v8::Local<v8::Value>&)’
   baton->callback = v8::Persistent<v8::Value>::New(callback);
                                                            ^
../src/serialport.cpp:52:60: note: candidate is:
In file included from /home/pi/.node-gyp/4.2.1/include/node/node.h:42:0,
                 from ../src/serialport.h:5,
                 from ../src/serialport.cpp:3:
/home/pi/.node-gyp/4.2.1/include/node/v8.h:7181:4: note: static T* v8::PersistentBase<T>::New(v8::Isolate*, T*) [with T = v8::Value]
 T* PersistentBase<T>::New(Isolate* isolate, T* that) {
    ^
/home/pi/.node-gyp/4.2.1/include/node/v8.h:7181:4: note:   candidate expects 2 arguments, 1 provided
../src/serialport.cpp:53:69: error: ‘New’ is not a member of ‘v8::String’
   baton->dataCallback = v8::Persistent<v8::Value>::New(options->Get(v8::String::New("dataCallback")));
                                                                     ^
../src/serialport.cpp:54:77: error: ‘New’ is not a member of ‘v8::String’
   baton->disconnectedCallback = v8::Persistent<v8::Value>::New(options->Get(v8::String::New("disconnectedCallback")));
                                                                             ^
../src/serialport.cpp:55:70: error: ‘New’ is not a member of ‘v8::String’
   baton->errorCallback = v8::Persistent<v8::Value>::New(options->Get(v8::String::New("errorCallback")));
                                                                      ^
../src/serialport.cpp:57:3: error: ‘uv_work_t’ was not declared in this scope
   uv_work_t* req = new uv_work_t();
   ^
../src/serialport.cpp:57:14: error: ‘req’ was not declared in this scope
   uv_work_t* req = new uv_work_t();
              ^
../src/serialport.cpp:57:24: error: expected type-specifier before ‘uv_work_t’
   uv_work_t* req = new uv_work_t();
                        ^
../src/serialport.cpp:59:33: error: ‘uv_default_loop’ was not declared in this scope
   uv_queue_work(uv_default_loop(), req, EIO_Open, (uv_after_work_cb)EIO_AfterOpen);
                                 ^
../src/serialport.cpp:59:41: error: ‘EIO_Open’ was not declared in this scope
   uv_queue_work(uv_default_loop(), req, EIO_Open, (uv_after_work_cb)EIO_AfterOpen);
                                         ^
../src/serialport.cpp:59:52: error: ‘uv_after_work_cb’ was not declared in this scope
   uv_queue_work(uv_default_loop(), req, EIO_Open, (uv_after_work_cb)EIO_AfterOpen);
                                                    ^
../src/serialport.cpp:59:82: error: ‘uv_queue_work’ was not declared in this scope
   uv_queue_work(uv_default_loop(), req, EIO_Open, (uv_after_work_cb)EIO_AfterOpen);
                                                                                  ^
../src/serialport.cpp:61:16: error: ‘class v8::HandleScope’ has no member named ‘Close’
   return scope.Close(v8::Undefined());
                ^
../src/serialport.cpp:61:36: error: too few arguments to function ‘v8::Local<v8::Primitive> v8::Undefined(v8::Isolate*)’
   return scope.Close(v8::Undefined());
                                    ^
In file included from /home/pi/.node-gyp/4.2.1/include/node/node.h:42:0,
                 from ../src/serialport.h:5,
                 from ../src/serialport.cpp:3:
/home/pi/.node-gyp/4.2.1/include/node/v8.h:315:27: note: declared here
   friend Local<Primitive> Undefined(Isolate* isolate);
                           ^
../src/serialport.cpp: At global scope:
../src/serialport.cpp:64:20: error: variable or field ‘EIO_AfterOpen’ declared void
 void EIO_AfterOpen(uv_work_t* req) {
                    ^
../src/serialport.cpp:64:20: error: ‘uv_work_t’ was not declared in this scope
../src/serialport.cpp:64:31: error: ‘req’ was not declared in this scope
 void EIO_AfterOpen(uv_work_t* req) {
                               ^
../src/serialport.cpp: In function ‘v8::Handle<v8::Value> Open(const int&)’:
../src/serialport.cpp:62:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
serialport.target.mk:88: recipe for target 'Release/obj.target/serialport/src/serialport.o' failed
make: *** [Release/obj.target/serialport/src/serialport.o] Error 1
make: Leaving directory '/home/pi/node_modules/gps/node_modules/serialport/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:270:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Linux 4.4.8+
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/pi/node_modules/gps/node_modules/serialport
gyp ERR! node -v v4.2.1
gyp ERR! node-gyp -v v3.0.3
gyp ERR! not ok 
npm ERR! Linux 4.4.8+
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "update"
npm ERR! node v4.2.1
npm ERR! npm  v2.14.7
npm ERR! code ELIFECYCLE

npm ERR! serialport@1.2.5 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the serialport@1.2.5 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the serialport package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls serialport
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/pi/npm-debug.log
```

**Comment by fivdi** [comment]
#833 was about the installation of serialport@3.1.2. serialport@3.1.2 is currently the latest non-beta version of serialport and it works fine with Node.js v4.2.1 (which is the version of Node.js for this issue  and #833.) However, this issue is being reported for serialport@1.2.5. serialport@1.2.5 was released many years ago and doesn't support Node.js v4.2.1. The `gps` package which is being installed here appears to depend on serialport@1.2.5, so it can't function correctly with Node.js v4.1.2.

**Comment by infusion** [comment]
But why does this problem occur only for updates? The dependency of `gps` works fine for an installation and should use any version (preferable the newest).

**Comment by fivdi** [comment]
The information returned by `npm info gps` shows that gps@0.1.0, which is the highest version of `gps`, was released on '2013-10-04T16:29:15.812Z', which was a long time ago. On the other hand, gps@0.0.10 was released on '2016-06-09T15:16:20.695Z', which was yesterday, but there is already the higher version v0.1.0 from 2013.

[npm update](https://docs.npmjs.com/cli/update) will update all the packages to the latest version respecting semver. To me, this means it will attempt to install gps@0.1.0. 

The package.json file for gps@0.1.0 contains the following dependencies which means it depends on serialport@1.2.5:

```
  "dependencies": {
    "serialport": "~1.2.2"
  }
```

Given that you're the author of `gps`, I'd suggest publishing a new version on npm with a version number higher than v0.1.0. Then try everything again to see if it works.

```
pi@raspberrypi:~ $ npm info gps
 
{ name: 'gps',
  description: 'A GPS NMEA parser library',
  'dist-tags': { latest: '0.0.10' },
  versions: [ '0.0.6', '0.0.7', '0.0.8', '0.0.9', '0.0.10', '0.1.0' ],
  maintainers: 
   [ 'andrewnez <andrewnez@gmail.com>',
     'infusion <robert@xarg.org>' ],
  time: 
   { modified: '2016-06-09T15:16:20.695Z',
     created: '2013-10-04T16:29:12.642Z',
     '0.1.0': '2013-10-04T16:29:15.812Z',
     '0.0.6': '2016-03-28T12:11:09.457Z',
     '0.0.7': '2016-04-05T13:40:30.024Z',
     '0.0.8': '2016-04-07T17:14:40.122Z',
     '0.0.9': '2016-05-29T09:54:16.506Z',
     '0.0.10': '2016-06-09T15:16:20.695Z' },
  author: 'Robert Eisele <robert@xarg.org> (http://www.xarg.org/)',
  repository: { type: 'git', url: 'git://github.com/infusion/GPS.js.git' },
  homepage: 'https://github.com/infusion/GPS.js',
  keywords: 
   [ 'nmea',
     'gps',
     'serial',
     'parser',
     'distance',
     'geo',
     'location',
     'rmc',
     'gga',
     'gll',
     'gsa',
     'vtg',
     'gva' ],
  bugs: { url: 'https://github.com/infusion/GPS.js/issues' },
  readmeFilename: 'README.md',
  license: 'MIT OR GPL-2.0',
  title: 'gps.js',
  version: '0.0.10',
  main: 'gps',
  private: false,
  directories: { example: 'examples' },
  engines: { node: '*' },
  scripts: { test: 'mocha tests/*.js' },
  devDependencies: 
   { express: '*',
     'socket.io': '*',
     serialport: '*',
     angles: '*',
     mocha: '*',
     chai: '*',
     byline: '*' },
  gitHead: '6307aaa76f1ca05b00a05a013e60a74c44906e6b',
  dist: 
   { shasum: 'd0cc2fb3d65cc29cfacddcd48172b2d1ce6f0708',
     tarball: 'https://registry.npmjs.org/gps/-/gps-0.0.10.tgz' } }
```

**Comment by infusion** [comment]
Ouch! That explains a lot, thanks!
